### ISSUE TITLE
PALcolour / samplerates

### ISSUE THREAD

**Issue by techmindorg** [simoninns/ld-colourfilter#1]
Hi Simon,

Looks like you're making good progress. I'll try and add more comments on the code at some point, but an "urgent" thing I'd like to explain is about the sample rate as a multiple of Fsc.
"timesOne" or "timesTwo" makes no sense at all - this would be undersampled and it would be impossible to decode the colour, as well as lacking horizontal luma resolution.
"timesThree" would technically work, but "isn't a thing".
"timesFive, Six, Seven, Eight" would all technically work, but also aren't "a thing".

timesFour - sampled at 4xFsc **is** a thing because you get exactly 4 samples per subcarrier cycle, which is not only enough to decode the subcarrier phase and amplitude accurately (and without any aliasing complications), it also allows numerical simplifications in the decode, specifically the sine/cosine lookup tables (generated in lines 220-226 in your most recent check-in) can be simplified to -1, 1,1,-1, etc for the sine and 1,1,-1,-1, etc for the cosine - which then lends itself to hardware/FPGA implementation (with some change the to effective refAmpl scaling factor).

In the interest of keeping the code flexible and universal, I would suggest that it is better to keep the sampleRate generally independent of Fsc in the code, so allow it to be an arbitrary value that is not a multiple of Fsc (e.g. 20MHz) - and be careful not to inadvertently get an integer division in line 221.  To avoid having to have magic PAL numbers outside the PAL part, you might want to  set up a function that sets sampleRate to 4Fsc (and possibly switches in 4Fsc optimised code)?

On my point about universality, while mainstream PAL is 4.43MHz, at 625 line (e.g. UK standard PAL-I), you can also get some obscure things that it's fun to be able to decode. I've come across analog HD-resolution security cameras which use PAL-like colour coding but at Fsc around 35 to 40MHz!

At some point, there is some thought to be given as to whether you want to go down the route of highly optimised (potentially real-time) decoding specifically at 4Fsc, or whether you want to keep fully-flexible decoding, that runs slower.

**Comment by simoninns** [comment]
Thanks Andrew; I'll make some more changes.  Right now I'm in the process of refactoring the application into a more generic application so I can merge in the existing NTSC comb-filter too - effectively making the 'decode' parts sub-modules of the application, so I can add in a variety of decoders including your own (hence the name change to ld-colourfilter).

I'm also specialising the application to the ld-decode project as the range of possible options for a generic PAL filter is just too wide and will confuse 'normal' users :)  So I've already fixed the app to 4xFsc as that's all that ld-decode provides; so the bug you point out shouldn't break anything in the short-term :)

Your original code is now class encapsulated into the PalColour class; this has been refactored to accept a Qt QImage image container class and outputs another QImage.  There are now get/set methods that control the configuration too, so all the decoding details are contained in the one class.  The upshot is that you can pull this class out and reuse it.  I've also fixed (what looked like) a scaling error in the original code (the 'norm' value was out by a factor of 64 iirc causing the colour killer to be perma-on which you'd then compensated for elsewhere - if I understood it all right!) and that enabled me to scale all the floats up to 32 bit and recode for scaled integers (which greatly increased the speed).  Since my knowledge of your algorithm is a bit limited my modifications might not be optimal - but we can smooth that out.

When you've got time drop me an email; perhaps we can skype and do a walk-through of the code.  On another note, please make sure you're happy with the attribution I've given you in the code - let me know if you'd like something changed.

**Comment by techmindorg** [comment]
I would suggest that an absolute bare minimum value for the sample rate is about 10MHz, though anything much less than around 13MHz is strongly inadvised and is liable to result in some degree of aliassing artifacts, including stripeiness in bands of colour saturation and residual dots.
For standard-definition PAL, there's not much point in going above about 20MHz sampling rate (and the computational effort will scale with something between the sample rate and the square of the sample rate), but it wouldn't be insane to process at 25MHz or something if that made sense for other parts of the processing chain (if you were processing the RF off 2" Quad tape ? The luminance bandwidth (horizontal resolution) of the late 1970's professional recordings was significantly more than the broadcast 6MHz!). An upper bound sanity-check of around 25MHz is reasonable, but put a comment in the code this isn't a hard limit, just that it takes increasingly more computation effort for negligible benefit as the rate increases.

Fsc should only really be 4.43361875 MHz (yes, that's strictly non-integer Hz) for regular PAL (e.g. UK PAL-I), or some values near to 3.58MHz for PAL-M or PAL-N. See for example https://www.datapro.net/techinfo/pal.html (as the first reference I found)
(For digital processing, as long as Fsc is specified to enough digits that the phase drift is insignificant over one scanline, that's good enough, hence why I specified it in round Hz)
If you just want to sanity-check inputs for SD resolution video, I'd say 3MHz <= Fsc <= 5MHz ... but as I mentioned before, I've seen proprietary HD "PAL" with Fsc up at 35-40MHz (which you'd have to sample at 100MHz or thereabouts). Put comments to that effect alongside the limits in the code.
